app/src/ui/diff/diff-helpers.tsx

```diff
@@ -352,3 +352,32 @@ export function canSelect(
 ): file is WorkingDirectoryFileChange {
   return file instanceof WorkingDirectoryFileChange
 }
+
+/** Gets the width in pixels of the diff line number gutter based on the number of digits in the number */
+export function getLineWidthFromDigitCount(digitAmount: number): number {
+  let diffSize: number
+
+  if (digitAmount <= 3) {
+    diffSize = 35
+  } else {
+    diffSize = 10 * digitAmount + 5
+  }
+
+  return diffSize
+}
+
+/** Utility function for getting the digit count of the largest line number in an array of diff hunks */
+export function getLargestLineNumberDigitCount(
+  hunks: DiffHunk[],
+  fromExpansion: boolean = false
+): number {
+  const finalHunkIndex =
+    fromExpansion && hunks.length > 2 ? hunks.length - 2 : hunks.length - 1
+  const lastLine =
+    hunks[finalHunkIndex].lines[hunks[finalHunkIndex].lines.length - 1]
+  const largestLine =
+    ((lastLine.newLineNumber ?? 0) > (lastLine.oldLineNumber ?? 0)
+      ? lastLine.newLineNumber
+      : lastLine.oldLineNumber) ?? 0
+  return largestLine.toString().length
+}
```

Comment:
I think the logic when `fromExpansion === true` is wrong 🤔 

If I understand correctly the logic around `hunks.length > 2 ? hunks.length - 2 : hunks.length - 1` is to avoid the dummy hunk used for expansion at the end of the diff, however that might not exist if the last hunk of the (real) diff is already located at the end.

A more generic approach to this would be a loop from the last line of the last hunk, going backwards, until you find a line that is not of `Hunk` type (and therefore has either `newLineNumber`, or `oldLineNumber`, or both). Sounds a bit excessive, but I think it's cleaner than having to take into account, explicitly, if the last hunk is one of these dummy hunks and/or if there are more than 2 or 1 hunks. Also, I think the intent of the resulting code would be clearer too 😄

Reply:
Also, given how cheap this implementation seems to be… maybe there is no need to do it on both expansion and parsing and we can just do it when needed as you had it before?

Sorry about that change of mind 😅  To be honest, when I suggested to do this on expansion and parsing, I had something different in mind: we would change the max line number as we expand a hunk or parse the diff line by line, instead of looking at the last line of the resulting diff after parsing/expanding.

If this is fast enough, I'd rather remove it from the expansion/parsing code 😳

Reply:
If you'll take a look I changed the logic for grabbing the last line number and parsing hunks. It seems to work alright and feels a lot cleaner than what I was doing before since it had a bit of code smell to me when I was skipping the last hunk or two to dodge the dummy.

As for cheapness I have to say that putting the `largestLineNumber` in the diff data and calculating it when parsing them feels a lot cleaner as the data feels like it belongs there. It also only runs once when the diffs load (sometimes for 2 different diffs when first loading, but I'm assuming that's when it's parsing every hunk in that diff) and then only when expanded. 

I personally think the approach you originally suggested seems a lot cleaner and efficient than what I was doing before as the logic doesn't seem to belong in the individual diffs and also replicates code a bit.

That's just my 2 cents though, so after looking at it, if you still want to swap it back I can.